### PR TITLE
fix(platform): smart truncate pr body

### PR DESCRIPTION
**Pull request by viceice**
see #2569, #4427

### CODE REVIEW

**Comment by ViceIce** [comment]
@rarkins Should we add this to all platforms?

**Comment by rarkins** [comment]
Yes

**Comment by ViceIce** [comment]
Done

**Comment by rarkins** [comment]
@ViceIce how did you verify the respective limits? I see that GitHub's went way up, while the others are very small ion comparison.

**Comment by ViceIce** [comment]
@rarkins I just reused exiting limits already in code

**Comment by rarkins** [comment]
I thought we were missing limits for some platforms though?

**Comment by ViceIce** [comment]
As i've checked the other platform i found that all except gitlab already have limits.

So i only added the smart truncate function too all platforms respecting their current limits.

File: lib/platform/azure/index.ts [hunk 1]
```diff
@@ -487,7 +488,7 @@ export async function mergePr(pr: number) {
 
 export function getPrBody(input: string) {
   // Remove any HTML we use
-  return input
+  return smartTruncate(input, 4000)
```
Comment (rarkins) [inline lib/platform/azure/index.ts]: 4000 seems low compared to the others

File: lib/platform/utils/pr-body.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import { appName } from '../../config/app-strings';
+
+export function smartTruncate(input: string, len = 60000): string {
```
Comment (rarkins) [inline lib/platform/utils/pr-body.ts]: I added 60000 to GitHub so I think it's best if we remove the default here

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/platform/azure/index.ts (hunk 1 above)
Comment (ViceIce) [reply lib/platform/azure/index.ts]: got it from here
https://github.com/renovatebot/renovate/blob/ee305b1a1719981684980399535789dbfe0494a2/lib/platform/azure/azure-helper.ts#L189-L194

Comment (rarkins) [reply lib/platform/azure/index.ts]: it checks out

File: lib/platform/utils/pr-body.ts (hunk 2 above)
Comment (ViceIce) [reply lib/platform/utils/pr-body.ts]: done

File: lib/platform/utils/pr-body.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  }
+  const releaseNotesMatch = input.match(
+    new RegExp(`### Release Notes.*### ${appName} configuration`, 'ms')
+  );
+  if (releaseNotesMatch) {
+    const divider = `</details>\n\n---\n\n### ${appName} configuration`;
+    const [releaseNotes] = releaseNotesMatch;
+    const nonReleaseNotesLength =
+      input.length - releaseNotes.length - divider.length;
+    const availableLength = len - nonReleaseNotesLength;
```
Comment (ViceIce) [inline lib/platform/utils/pr-body.ts]: @rarkins Should we catch possible negative `availableLength` ?

Comment (rarkins) [reply lib/platform/utils/pr-body.ts]: Yes, good idea

File: lib/platform/azure/index.ts (hunk 1 above)
Comment (ViceIce) [reply lib/platform/azure/index.ts]: https://github.com/renovatebot/renovate/blob/8554df5c612ba5035c4f9b7061a4b45ca59b3dbb/lib/platform/azure/index.ts#L357

Comment (ViceIce) [reply lib/platform/azure/index.ts]: https://developercommunityapi.westus.cloudapp.azure.com/content/problem/608770/remove-4000-character-limit-on-pull-request-descri.html

File: lib/platform/utils/pr-body.ts (hunk 3 above)
Comment (ViceIce) [reply lib/platform/utils/pr-body.ts]: What should we do then? No truncate or fallback to simple substring?

Comment (rarkins) [reply lib/platform/utils/pr-body.ts]: I think fall back to substring. It's a pretty hopeless case anyway!

Comment (ViceIce) [reply lib/platform/utils/pr-body.ts]: done

**Pull request merged**